oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java

```diff
@@ -142,14 +143,22 @@ public void validateAuthorizationCode(String code) {
     public static AbstractJwsSigner createJwsSigner(Jwt idToken, OpenIdConfigurationResponse discoveryResponse, PublicOpKeyService keyService, OpClientFactory opClientFactory, Rp rp) {
         final String algorithm = idToken.getHeader().getClaimAsString(JwtHeaderName.ALGORITHM);
         final SignatureAlgorithm signatureAlgorithm = SignatureAlgorithm.fromString(algorithm);
-        final String kid = idToken.getHeader().getClaimAsString(JwtHeaderName.KEY_ID);
         final String jwkUrl = discoveryResponse.getJwksUri();
+        String kid = idToken.getHeader().getClaimAsString(JwtHeaderName.KEY_ID);
 
         if (signatureAlgorithm == null)
             throw new HttpException(ErrorResponseCode.INVALID_ALGORITHM);
 
+        if (Strings.isNullOrEmpty(kid) && (signatureAlgorithm.getFamily() == AlgorithmFamily.RSA || signatureAlgorithm.getFamily() == AlgorithmFamily.EC)) {
+            LOG.warn("`kid` is missing in id_token. Checking keystore for `kid`");
+            kid = keyService.getKeyId(idToken, jwkUrl, signatureAlgorithm, Use.SIGNATURE);
```

Comment:
I'm not sure we need this extra `getKeyId` method. If we call `keyService.getRSAPublicKey()` and it returns result then we have it. If it throws exception then we don't have it and in addition it's faster since for existing `kid` it's already cached.

Reply:
as discussed on chat, we need `getKeyId` method to pass the conformance test.
We have created `getPublicKey` method in `keyService` to get keys for `ECDSA` and `RSA`.